crypto/multisig/src/v2.rs

```diff
@@ -0,0 +1,1260 @@
+    /// their byte representation since we always sort them). Here is an
+    /// example that demonstrates the implications of this: Assume a signer set,
+    /// that requires 1-of-2 and is sorted, and where each member is a
+    /// m-of-n group where the first requires 2/3 and the second requires
+    /// 3/3:
+    /// 1) A, B, C (threshold 2)
+    /// 2) A, D, E (threshold 3)
+    /// If we are passed signatures A, B, C, D, E we could in theory satisfy the
+    /// threshold of both groups but since A gets "consumed" when matching the
+    /// first group, leaving the second group with only D and E, it will not
+    /// be considered a match. We say we could match all 5 signatures
+    /// in theory, because if we ignored A for the first set, it could still be
+    /// used to match the 2nd set. Such algorithm would require a more
+    /// complex (and slower) implementation, and is not currently supported.
+    pub fn verify<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<SIG>,
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
+    {
+        if multi_sig.signatures().len() > MAX_SIGNATURES {
+            return Err(SignatureError::new());
+        }
+
+        // Sort and dedup the list of signatures..
+        let mut signatures = multi_sig.signatures.clone();
+        signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
+        signatures.dedup();
+
+        // Verify signatures.
+        self.verify_helper(message, &signatures, &[])
+    }
+
+    fn verify_helper<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        signatures: &[SIG],
+        seen_signers: &[S],
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
+    {
+        // Sort and dedup the list of signers.
+        // While the verification code below should be immune to duplicate signers, the
+        // overhead of deduping them is negligible and being extra-safe is a good idea.
+        let mut signers = self.signers.clone();
+        signers.sort();
+        signers.dedup();
+
+        // See which signers are satisfied by which signatures.
+        let mut matched_signers = Vec::new();
+        let mut num_matched_entities = 0;
+
+        for signer in signers.iter() {
+            match signer.entity {
+                Some(SignerEntity::Single(ref s)) => {
+                    // If we already encountered this signer, we cannot use it again.
```

Comment:
I don't understand right now why we are concerned about this.

1. Didn't we remove duplicate signers in the `signers.dedup()` call?
2. Why is it bad to have a signer appear twice in a set? Is this not like giving them two votes? I don't see anything obviously wrong with that.

Reply:
We remove signers in the dedup but that doesn't deal with cases where the signer appears twice in different signersets. e.g. you can have a signer set S1 that has [A, B], and then S2 that has [A, C] and S3 that has [S1, S2] and deduping on S3 will still result in signer A appearing twice, and if we find two different signatures signed by A (because they figured out a way to produce two signatures that are valid for the same data but look differently) then we would potentially count them twice, and that is undesirable.
So the approach here is better safe than sorry - we want to be absolutely sure they don't get two votes, so that's the thinking here.

Reply:
I think the question of "was signer set X satisfied by the given set of signatures" should be totally independent to the question "was signer set Y satisfied by the given set of signatures". It is extremely complicated to do otherwise, and it will mean that if e.g. a non-technical person wants to understand what the significance will be in practice of adding an extra nested signer group to a signing config, they simply won't be able to do it successfully if there are overlapping signer sets.

So I think this "extra safety" is a defect rather than a feature.

Reply:
If we want to have this safety, IMO there should be like a first signer-set validation step that just scans across the signer set and if a signer appears twice, rejects the signer set without looking at any signatures. That would allow to simplify the signature checking code, and to give a better error message in this case.

But again, I think this check is pointless and shouldn't be done at all.

Reply:
I'm fine simplifying it to only making sure the same signer is not used twice in the same signer set, and not worrying about the signer appearing multiple times in different sets.